File: src/util/repo.ts
import { spawn } from "node:child_process";
import { mkdir, readFile, readdir, stat } from "node:fs/promises";
import { join, relative } from "node:path";
import type { RepoContext } from "../types";
import { log } from "./log";

/** Stage 0 — Repo ingest. Clones/updates the repo and gathers the file tree +
 * metadata. Deliberately does NOT pre-filter "key files" or excerpt them — codex
 * reads the real files itself during analyze/lesson stages. */

const IGNORED_DIRS = new Set([
  "node_modules", ".git", "dist", "build", "out", ".next", "target",
  ".repo2learn", "coverage", ".turbo", ".cache", "vendor", ".venv", "venv",
]);

export async function ingestRepo(repo: string, workDir: string): Promise<RepoContext> {
  const isUrl = /^https?:\/\//.test(repo) || /^git@/.test(repo);
  const localPath = isUrl ? await cloneOrPull(repo, workDir) : await resolveLocal(repo);
  const name = deriveName(repo, localPath);
  log.info(`ingesting ${name} from ${localPath}`);
  const tree = await collectTree(localPath);
  const { languages, loc } = await analyze(localPath, tree);
  const sha = await gitSha(localPath);
  const summary = await readReadme(localPath);
  log.ok(`ingest done · ${tree.length} files · ${loc} LOC · ${Object.keys(languages).length} languages`);
  return { url: repo, localPath, sha, name, defaultBranch: "main", summary, loc, languages, tree };
}

async function cloneOrPull(url: string, workDir: string): Promise<string> {
  await mkdir(workDir, { recursive: true });
  const dir = join(workDir, dirNameForUrl(url));
  const exists = await safeStat(join(dir, ".git"));
  if (exists) { await runGit(["pull", "--ff-only"], dir); return dir; }
  await runGit(["clone", "--depth", "1", url, dir], workDir);
  return dir;
}
async function resolveLocal(path: string): Promise<string> {
  const st = await safeStat(path);
  if (!st || !st.isDirectory()) throw new Error(`repo path not found or not a directory: ${path}`);
  return path;
}
async function collectTree(root: string): Promise<string[]> {
  const out: string[] = [];
  async function walk(dir: string): Promise<void> {
    const entries = await readdir(dir, { withFileTypes: true });
    for (const e of entries) {
      if (e.name.startsWith(".") && e.name !== ".") continue;
      const full = join(dir, e.name);
      if (e.isDirectory()) { if (!IGNORED_DIRS.has(e.name)) await walk(full); }
      else if (e.isFile()) { out.push(relative(root, full)); }
    }
  }
  await walk(root);
  return out.sort();
}
function langOf(file: string): string {
  const ext = file.slice(file.lastIndexOf(".")).toLowerCase();
  const map: Record<string, string> = {
    ".ts":"TypeScript",".tsx":"TypeScript",".js":"JavaScript",".jsx":"JavaScript",".mjs":"JavaScript",".cjs":"JavaScript",
    ".py":"Python",".go":"Go",".rs":"Rust",".java":"Java",".kt":"Kotlin",".rb":"Ruby",".php":"PHP",".c":"C",".h":"C",
    ".cpp":"C++",".cs":"C#",".swift":"Swift",".md":"Markdown",".mdx":"MDX",".css":"CSS",".scss":"SCSS",".html":"HTML",
    ".json":"JSON",".yml":"YAML",".yaml":"YAML",".toml":"TOML",".sh":"Shell",
  };
  return map[ext] ?? "Other";
}
async function analyze(root: string, tree: string[]) {
  const bytes: Record<string, number> = {}; let loc = 0;
  const CODE = new Set([".ts",".tsx",".js",".jsx",".mjs",".cjs",".py",".go",".rs",".java",".kt",".rb",".c",".h",".cpp",".cs",".swift"]);
  for (const rel of tree) {
    const ext = rel.slice(rel.lastIndexOf(".")).toLowerCase();
    try {
      const content = await readFile(join(root, rel), "utf8");
      bytes[langOf(rel)] = (bytes[langOf(rel)] ?? 0) + content.length;
      if (CODE.has(ext)) loc += content.split("\n").filter((l) => l.trim().length > 0).length;
    } catch {}
  }
  const total = Object.values(bytes).reduce((a, b) => a + b, 0) || 1;
  const languages: Record<string, number> = {};
  for (const [k, v] of Object.entries(bytes)) languages[k] = +(v / total).toFixed(3);
  return { languages, loc };
}
async function readReadme(root: string): Promise<string> {
  for (const name of ["README.md", "readme.md", "README.MD", "README"]) {
    try { return (await readFile(join(root, name), "utf8")).slice(0, 800).replace(/\s+/g, " ").trim(); } catch {}
  }
  return "";
}
async function gitSha(root: string): Promise<string> {
  try { return (await runGit(["rev-parse", "--short", "HEAD"], root)).trim() || "unknown"; } catch { return "unknown"; }
}
function runGit(args: string[], cwd: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const p = spawn("git", args, { cwd, stdio: ["ignore", "pipe", "pipe"] });
    const out: string[] = [], err: string[] = [];
    p.stdout.on("data", (d) => out.push(d.toString())); p.stderr.on("data", (d) => err.push(d.toString()));
    p.on("error", reject); p.on("close", (c) => c === 0 ? resolve(out.join("")) : reject(new Error(err.join("").slice(0, 300))));
  });
}
const safeStat = (p: string) => stat(p).catch(() => undefined);
function deriveName(repo: string, localPath: string): string {
  const base = repo.split(/[\/:]/).filter(Boolean).pop() ?? localPath.split(/[\\/]/).pop() ?? "repo";
  return base.replace(/\.git$/, "");
}
/** Directory name a repo URL is cloned into under workDir. Exported so the
 *  server can locate (and clean up) a clone without re-deriving the rule. */
export function dirNameForUrl(url: string): string {
  const base = url.split(/[\/:]/).filter(Boolean).pop() ?? "repo";
  return base.replace(/\.git$/, "").replace(/[^a-z0-9_-]+/gi, "-");
}

File: backend/app/services/repo.py
from __future__ import annotations

import asyncio
from pathlib import Path

from pydantic import BaseModel

from app.core.config import Settings, get_settings
from app.services.store import dir_name_for_url

IGNORED_DIRS = {
    "node_modules",
    ".git",
    "dist",
    "build",
    "out",
    ".next",
    "target",
    ".repo2learn",
    "coverage",
    ".turbo",
    ".cache",
    "vendor",
    ".venv",
    "venv",
}

CODE_EXTENSIONS = {
    ".ts",
    ".tsx",
    ".js",
    ".jsx",
    ".mjs",
    ".cjs",
    ".py",
    ".go",
    ".rs",
    ".java",
    ".kt",
    ".rb",
    ".c",
    ".h",
    ".cpp",
    ".cs",
    ".swift",
}


class RepoContext(BaseModel):
    url: str
    localPath: str
    sha: str
    name: str
    defaultBranch: str
    summary: str
    loc: int
    languages: dict[str, float]
    tree: list[str]


async def ingest_repo(repo: str, settings: Settings | None = None) -> RepoContext:
    cfg = settings or get_settings()
    is_url = repo.startswith(("http://", "https://", "git@"))
    local_path = await _clone_or_pull(repo, cfg.work_dir) if is_url else Path(repo)
    if not local_path.is_dir():
        raise FileNotFoundError(f"repo path not found or not a directory: {repo}")

    tree = _collect_tree(local_path)
    languages, loc = _analyze(local_path, tree)
    return RepoContext(
        url=repo,
        localPath=str(local_path),
        sha=await _git(["rev-parse", "--short", "HEAD"], local_path, fallback="unknown"),
        name=_derive_name(repo, local_path),
        defaultBranch=await _git(
            ["rev-parse", "--abbrev-ref", "HEAD"],
            local_path,
            fallback="main",
        ),
        summary=_read_readme(local_path),
        loc=loc,
        languages=languages,
        tree=tree,
    )


async def _clone_or_pull(url: str, work_dir: Path) -> Path:
    work_dir.mkdir(parents=True, exist_ok=True)
    target = work_dir / dir_name_for_url(url)
    if (target / ".git").exists():
        await _git(["pull", "--ff-only"], target)
        return target
    await _git(["clone", "--depth", "1", url, str(target)], work_dir)
    return target


async def _git(args: list[str], cwd: Path, fallback: str | None = None) -> str:
    proc = await asyncio.create_subprocess_exec(
        "git",
        *args,
        cwd=str(cwd),
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    stdout, stderr = await proc.communicate()
    if proc.returncode == 0:
        return stdout.decode("utf-8", errors="replace").strip()
    if fallback is not None:
        return fallback
    raise RuntimeError(stderr.decode("utf-8", errors="replace")[-800:])


def _collect_tree(root: Path) -> list[str]:
    out: list[str] = []
    for path in root.rglob("*"):
        rel_parts = path.relative_to(root).parts
        if any(part in IGNORED_DIRS for part in rel_parts):
            continue
        if path.is_file():
            out.append(path.relative_to(root).as_posix())
    return sorted(out)


def _analyze(root: Path, tree: list[str]) -> tuple[dict[str, float], int]:
    bytes_by_lang: dict[str, int] = {}
    loc = 0
    for rel in tree:
        path = root / rel
        try:
            content = path.read_text(encoding="utf-8")
        except UnicodeDecodeError:
            continue
        except OSError:
            continue
        lang = _language_of(path)
        bytes_by_lang[lang] = bytes_by_lang.get(lang, 0) + len(content)
        if path.suffix.lower() in CODE_EXTENSIONS:
            loc += sum(1 for line in content.splitlines() if line.strip())

    total = sum(bytes_by_lang.values()) or 1
    return {lang: round(size / total, 3) for lang, size in bytes_by_lang.items()}, loc


def _language_of(path: Path) -> str:
    return {
        ".ts": "TypeScript",
        ".tsx": "TypeScript",
        ".js": "JavaScript",
        ".jsx": "JavaScript",
        ".mjs": "JavaScript",
        ".cjs": "JavaScript",
        ".py": "Python",
        ".go": "Go",
        ".rs": "Rust",
        ".java": "Java",
        ".kt": "Kotlin",
        ".rb": "Ruby",
        ".php": "PHP",
        ".c": "C",
        ".h": "C",
        ".cpp": "C++",
        ".cs": "C#",
        ".swift": "Swift",
        ".md": "Markdown",
        ".mdx": "MDX",
        ".css": "CSS",
        ".scss": "SCSS",
        ".html": "HTML",
        ".json": "JSON",
        ".yml": "YAML",
        ".yaml": "YAML",
        ".toml": "TOML",
        ".sh": "Shell",
    }.get(path.suffix.lower(), "Other")


def _read_readme(root: Path) -> str:
    for name in ["README.md", "readme.md", "README.MD", "README"]:
        try:
            return " ".join((root / name).read_text(encoding="utf-8").split())[:800]
        except OSError:
            continue
    return ""


def _derive_name(repo: str, local_path: Path) -> str:
    base = repo.replace("\\", "/").rstrip("/").split("/")[-1] or local_path.name
    return base.removesuffix(".git")

